File: timerx-core/src/main/java/timerx/TimerImpl.java
package timerx;

import static timerx.TimeCountingState.INACTIVE;
import static timerx.TimeCountingState.PAUSED;
import static timerx.TimeCountingState.RESUMED;

import android.annotation.SuppressLint;
import android.os.Handler;
import android.os.Message;
import android.os.SystemClock;
import androidx.annotation.NonNull;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.concurrent.TimeUnit;

class TimerImpl implements Timer {

  private final int MSG = 3;

  // Start time in milliseconds
  private final long startTime;

  // Current time (left time) in millis
  private long currentTime;

  // Millis when timer should stop
  private long millisInFuture;

  // Interval for timer
  private long interval;

  private TimeTickListener tickListener;
  private final Action finishAction;
  private final Semantic startSemantic;

  private TimeCountingState state = INACTIVE;
  private StringBuilderTimeFormatter timeFormatter;

  private final SortedSet<NextFormatsHolder> nextFormatsHolders;
  private final SortedSet<ActionsHolder> nextActionsHolders;
  private SortedSet<NextFormatsHolder> copyOfFormatsHolders;
  private SortedSet<ActionsHolder> copyOfNextActionsHolders;

  TimerImpl(long startTime, Semantic startSemantic, TimeTickListener tickListener,
      Action finishAction, SortedSet<NextFormatsHolder> nextFormatsHolders,
      SortedSet<ActionsHolder> nextActionsHolders) {
    this.startTime = startTime;
    this.startSemantic = startSemantic;
    this.tickListener = tickListener;
    this.finishAction = finishAction;
    this.nextFormatsHolders = nextFormatsHolders;
    this.nextActionsHolders = nextActionsHolders;
    currentTime = startTime;
  }

  @NonNull
  @Override
  public CharSequence getFormattedStartTime() {
    return new StringBuilderTimeFormatter(startSemantic).format(startTime);
  }

  @Override
  public void start() {
    if (state != RESUMED) {
      if (state == INACTIVE) {
        currentTime = startTime;
        copyOfFormatsHolders = new TreeSet<>(nextFormatsHolders);
        copyOfNextActionsHolders = new TreeSet<>(nextActionsHolders);
        applyFormat(startSemantic);
        millisInFuture = SystemClock.elapsedRealtime() + startTime;
      } else {
        millisInFuture = SystemClock.elapsedRealtime() + currentTime;
      }
      handler.sendMessage(handler.obtainMessage(MSG));
      state = RESUMED;
    }
  }

  @Override
  public void stop() {
    state = PAUSED;
    handler.removeMessages(MSG);
  }

  @Override
  public void reset() {
    state = INACTIVE;
    handler.removeMessages(MSG);
    applyFormat(startSemantic);
  }

  @Override
  public long getRemainingTimeIn(@NonNull TimeUnit timeUnit) {
    return timeUnit.convert(currentTime, TimeUnit.MILLISECONDS);
  }

  @Override
  public void release() {
    nextFormatsHolders.clear();
    copyOfFormatsHolders.clear();
    nextActionsHolders.clear();
    copyOfNextActionsHolders.clear();
    tickListener = null;
    handler.removeCallbacksAndMessages(null);
    handler = null;
  }

  private void applyFormat(Semantic semantic) {
    timeFormatter = new StringBuilderTimeFormatter(semantic);
    interval = timeFormatter.getOptimalDelay();
  }

  @SuppressLint("HandlerLeak")
  private Handler handler = new Handler() {

    @Override
    public void handleMessage(Message msg) {
      synchronized (TimerImpl.this) {
        long startExecution = SystemClock.elapsedRealtime();
        currentTime = millisInFuture - SystemClock.elapsedRealtime();
        changeFormatIfNeed();
        makeActionIfNeed();
        if (currentTime <= 0) {
          finishTimer();
          return;
        }
        if (tickListener != null) {
          CharSequence formattedTime = timeFormatter.format(currentTime);
          tickListener.onTick(formattedTime);
        }
        long executionTime = SystemClock.elapsedRealtime() - startExecution;
        sendMessageDelayed(obtainMessage(MSG), interval - executionTime);
      }
    }
  };

  private void changeFormatIfNeed() {
    if (copyOfFormatsHolders.size() > 0
        && !timeFormatter.getFormat()
        .equals(copyOfFormatsHolders.first().getSemantic().getFormat())
        && currentTime <= copyOfFormatsHolders.first().getMillis()) {
      applyFormat(copyOfFormatsHolders.first().getSemantic());
      copyOfFormatsHolders.remove(copyOfFormatsHolders.first());
    }
  }

  private void makeActionIfNeed() {
    if (copyOfNextActionsHolders.size() > 0
        && currentTime <= copyOfNextActionsHolders.first().getMillis()) {
      copyOfNextActionsHolders.first().getAction().run();
      copyOfNextActionsHolders.remove(copyOfNextActionsHolders.first());
    }
  }

  private void finishTimer() {
    currentTime = 0;
    if (tickListener != null) {
      tickListener.onTick(timeFormatter.format(currentTime));
    }
    if (finishAction != null) {
      finishAction.run();
    }
    reset();
  }
}

File: timerx-core/src/main/java/timerx/TimerImpl.kt
package timerx

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Message
import android.os.SystemClock
import timerx.TimeCountingState.INACTIVE
import timerx.TimeCountingState.PAUSED
import timerx.TimeCountingState.RESUMED
import timerx.formatting.Constants.TimeValues
import timerx.formatting.Semantic
import timerx.formatting.TimeFormatterFactory
import timerx.formatting.TimeUnitType
import java.util.concurrent.TimeUnit
import kotlin.math.abs

internal class TimerImpl(
  private val useExactDelay: Boolean,
  private var tickListener: TimeTickListener?,
  private val finishAction: Runnable?,
  private val semantics: MutableList<SemanticHolder>,
  private val actions: MutableList<ActionsHolder>
) : Timer {
  
  // Remaining (current time) in millis
  private var remainingTime: Long = semantics.first().millis
  
  // Time in future when timer should stop
  private var millisInFuture: Long = TimeValues.NONE
  
  // Delay (in millis) for delaying timer
  private var currentDelay: Long = 0
  
  // Delay (in millis) for correcting timer ticking
  private var currentExactDelay: Long = 0
  
  // Handler for scheduling actions and formats change
  private val workerHandler = Handler()
  
  // State of the timer
  private var state = INACTIVE
  
  // Current time formatter
  private var timeFormatter = TimeFormatterFactory.create(semantics.first().semantic)
  
  // Mode when all formats don't have milliseconds
  private val isInSimpleSecondsMode = semantics.all {
    !it.semantic.has(TimeUnitType.R_MILLISECONDS)
  }
  
  override val formattedStartTime: CharSequence
    get() {
      val holder = semantics.first()
      return TimeFormatterFactory.create(holder.semantic).format(holder.millis)
    }
  
  override val remainingTimeInMillis: Long
    get() = remainingTime
  
  override val remainingFormattedTime: CharSequence
    get() = timeFormatter.format(remainingTime)
  
  override fun start() {
    if (state == RESUMED) return
    if (millisInFuture == TimeValues.NONE) {
      val startTime = semantics.first().millis
      remainingTime = startTime
      applyFormat(semantics.first().semantic)
      val timeToSkip = if (useExactDelay) currentExactDelay else 0
      millisInFuture = SystemClock.elapsedRealtime() + startTime - timeToSkip
    } else {
      val timeToSkip = if (useExactDelay) currentExactDelay else 0
      millisInFuture = SystemClock.elapsedRealtime() + remainingTime - timeToSkip
    }
    handler!!.sendMessage(handler!!.obtainMessage())
    scheduleFormatsAndActionsChange()
    state = RESUMED
  }
  
  override fun stop() {
    state = PAUSED
    handler!!.removeCallbacksAndMessages(null)
    cancelFormatsAndActionsChanges()
  }
  
  override fun setTime(time: Long, timeUnit: TimeUnit) {
    val newTime = timeUnit.toMillis(time)
    millisInFuture = SystemClock.elapsedRealtime() + newTime
    remainingTime = newTime
    cancelFormatsAndActionsChanges()
    applyAppropriateFormat()
    scheduleFormatsAndActionsChange()
  }
  
  override fun reset() {
    state = INACTIVE
    millisInFuture = TimeValues.NONE
    remainingTime = semantics.first().millis
    handler!!.removeCallbacksAndMessages(null)
    cancelFormatsAndActionsChanges()
    applyFormat(semantics.first().semantic)
  }
  
  override fun release() {
    semantics.clear()
    actions.clear()
    tickListener = null
    cancelFormatsAndActionsChanges()
    handler?.removeCallbacksAndMessages(null)
    handler = null
  }
  
  private fun applyFormat(semantic: Semantic) {
    synchronized(this) {
      timeFormatter = TimeFormatterFactory.create(semantic)
      currentDelay = timeFormatter.getWaitingDelay(useExactDelay)
      currentExactDelay = timeFormatter.getWaitingDelay(useExactDelay = true)
    }
  }
  
  @SuppressLint("HandlerLeak")
  private var handler: Handler? = object : Handler() {
    
    override fun handleMessage(msg: Message) {
      synchronized(this@TimerImpl) {
        val startExecution = SystemClock.elapsedRealtime()
        if (isInSimpleSecondsMode && useExactDelay) {
          remainingTime = remainingTime.coerceAtLeast(0)
          val oldRemainingTime = remainingTime
          tickListener?.onTick(remainingTime, timeFormatter.format(remainingTime))
          remainingTime = millisInFuture - SystemClock.elapsedRealtime()
          val expectedTime = oldRemainingTime - currentExactDelay
          val diff = abs(remainingTime - expectedTime)
          remainingTime += diff
          if (remainingTime <= 0) {
            postDelayed(::finishTimer, currentExactDelay)
            return
          }
        } else {
          remainingTime = millisInFuture - SystemClock.elapsedRealtime()
          remainingTime = remainingTime.coerceAtLeast(0)
          tickListener?.onTick(remainingTime, timeFormatter.format(remainingTime))
          if (remainingTime <= 0) {
            finishAction?.run()
            reset()
            return
          }
        }
        val executionTime = SystemClock.elapsedRealtime() - startExecution
        sendMessageDelayed(obtainMessage(), currentDelay - executionTime)
      }
    }
  }
  
  private fun applyAppropriateFormat() {
    if (remainingTime > semantics.first().millis) {
      applyFormat(semantics.first().semantic)
      return
    }
    for (i in semantics.indices.reversed()) {
      if (semantics[i].millis > remainingTime) {
        applyFormat(semantics[i].semantic)
        break
      }
    }
  }
  
  private fun scheduleFormatsAndActionsChange() {
    semantics.forEach { holder ->
      if (holder.millis < remainingTime) {
        workerHandler.postDelayed({ applyFormat(holder.semantic) },
          remainingTime - holder.millis)
      }
    }
    actions.forEach { holder ->
      if (holder.millis < remainingTime) {
        workerHandler.postDelayed({ holder.action.run() },
          remainingTime - holder.millis)
      }
    }
  }
  
  private fun cancelFormatsAndActionsChanges() {
    workerHandler.removeCallbacksAndMessages(null)
  }
  
  private fun finishTimer() {
    remainingTime = 0
    tickListener?.onTick(remainingTime, timeFormatter.format(remainingTime))
    finishAction?.run()
    reset()
  }
}
